File: ruby/source/diskstats.rb
require 'solanum/source'

class Solanum::Source::Diskstats < Solanum::Source
  attr_reader :devices

  STAT_FILE = '/proc/diskstats'

  FIELDS = %w{
    major minor name
    reads_completed reads_merged read_sectors read_time
    writes_completed writes_merged write_sectors write_time
    io_active io_time io_weighted_time
  }


  def initialize(opts)
    super(opts)
    @devices = opts['devices'] || []
    @detailed = opts['detailed'] || false
    @last = {}
  end


  def parse_stats(line)
    columns = line.strip.split(/\s+/)
    columns = columns.take(3) + columns.drop(3).map(&:to_i)
    Hash[FIELDS.zip(columns)]
  end


  def collect!
    events = []

    File.readlines(STAT_FILE).each do |line|
      stats = parse_stats(line)
      device = stats['name']

      if @devices.include?(device) || (@devices.empty? && device =~ /^sd[a-z]$/)
        if @last[device]
          diff = Hash.new(0)
          FIELDS.drop(3).each do |field|
            diff[field] = stats[field] - @last[device][field]
          end
          prefix = "diskstats #{device}"

          # Reads

          events << {
            service: "#{prefix} read sectors",
            metric: diff['read_sectors'],
          }

          events << {
            service: "#{prefix} read time",
            metric: diff['read_time'],
          }

          if @detailed
            events << {
              service: "#{prefix} read completed",
              metric: diff['reads_completed'],
            }

            events << {
              service: "#{prefix} read merged",
              metric: diff['reads_merged'],
            }
          end

          # Writes

          events << {
            service: "#{prefix} write sectors",
            metric: diff['write_sectors'],
          }

          events << {
            service: "#{prefix} write time",
            metric: diff['write_time'],
          }

          if @detailed
            events << {
              service: "#{prefix} write completed",
              metric: diff['writes_completed'],
            }

            events << {
              service: "#{prefix} write merged",
              metric: diff['writes_merged'],
            }
          end

          # IO

          events << {
            service: "#{prefix} io active",
            metric: diff['io_active'],
          }

          events << {
            service: "#{prefix} io time",
            metric: diff['io_time'],
          }

          events << {
            service: "#{prefix} io weighted-time",
            metric: diff['io_weighted_time'],
          }
        end
        @last[device] = stats
      end
    end

    events
  end

end

File: src/solanum/source/disk_stats.clj
(ns solanum.source.disk-stats
  "Metrics source that measures the IO utilization of a host's block devices."
  (:require
    [clojure.string :as str]
    [clojure.tools.logging :as log]
    [solanum.source.core :as source]
    [solanum.system.core :as sys]
    [solanum.system.linux :as linux]))


;; ## Measurements

(source/defsupport :disk-stats
  #{:linux})


(def ^:private linux-disk-fields
  "Fields in the diskstats proc file."
  [:reads-completed :reads-merged :read-sectors :read-time
   :writes-completed :writes-merged :write-sectors :write-time
   :io-active :io-time :io-weighted-time])


(defn- parse-disk-counters
  "Parse a line from `/proc/diskstats` into a tuple with the device name and
  a map of stat keys to numeric counter values."
  [line]
  (let [columns (str/split (str/trim line) #"\s+")
        [major minor device] (take 3 columns)
        counters (->> (drop 3 columns)
                      (map #(Long/parseLong %))
                      (zipmap linux-disk-fields))]
    [device counters]))


(defn- read-disk-stats
  "Read disk device counters from `/proc/diskstats`."
  []
  (into {}
        (map parse-disk-counters)
        (linux/read-proc-lines "/proc/diskstats")))


(defn- measure-linux
  "Measure disk IO on Linux systems by reading the proc subsystem."
  [tracker]
  (let [data (read-disk-stats)
        prev @tracker]
    (reset! tracker data)
    (when prev
      (source/diff-tracker prev data))))


;; ## Disk Stats Source

(defrecord DiskStatsSource
  [tracker devices detailed]

  source/Source

  (collect-events
    [this]
    (let [info (case (:mode this)
                 :linux (measure-linux tracker))]
      (into []
            (comp
              (if (seq devices)
                (filter (comp (set devices) key))
                (filter #(re-matches #"(sd|xvd)[a-z]" (key %))))
              (mapcat
                (fn expand-events
                  [[device diffs]]
                  (keep
                    (fn [[service stat detailed-metric f]]
                      (when-let [metric (and (or detailed (not detailed-metric))
                                             (get diffs stat))]
                        {:service (str "disk stats " service)
                         :metric (if f (f metric) metric)
                         :device device}))
                    [["read bytes" :read-sectors false (partial * 512)]
                     ["read time" :read-time]
                     ["read completed" :reads-completed true]
                     ["read merged" :reads-merged true]
                     ["write bytes" :write-sectors false (partial * 512)]
                     ["write time" :write-time]
                     ["write completed" :writes-completed true]
                     ["write merged" :writes-merged true]
                     ["io active" :io-active]
                     ["io time" :io-time]
                     ["io weighted-time" :io-weighted-time]]))))
            info))))


(defmethod source/initialize :disk-stats
  [config]
  (map->DiskStatsSource
    {:tracker (atom {})
     :devices (not-empty (set (:devices config)))
     :detailed (boolean (:detailed config))}))
